x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/downsample/RollupShardIndexer.java

```diff
@@ -186,20 +198,44 @@ private void checkCancelled() {
                 task.getNumIndexed(),
                 task.getNumFailed()
             );
+            task.setRollupShardIndexerStatus(RollupShardIndexerStatus.CANCELLED);
             throw new TaskCancelledException(format("Shard %s rollup cancelled", indexShard.shardId()));
         }
     }
 
     private BulkProcessor2 createBulkProcessor() {
         final BulkProcessor2.Listener listener = new BulkProcessor2.Listener() {
+            private long bulkStartTime;
+
             @Override
             public void beforeBulk(long executionId, BulkRequest request) {
                 task.addNumSent(request.numberOfActions());
+                task.setBeforeBulkInfo(
+                    new RollupBeforeBulkInfo(
+                        System.currentTimeMillis(),
+                        executionId,
+                        request.estimatedSizeInBytes(),
+                        request.numberOfActions()
+                    )
+                );
+                bulkStartTime = System.currentTimeMillis();
             }
 
             @Override
             public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+                long bulkDurationInMillis = System.currentTimeMillis() - bulkStartTime;
```

Comment:
This can go negative. We should either protect by wrapping in `Math.max(1, System.currentTimeMillis() - bulkStartTime)` or base everything on `System.nanoTime()`. I think we should go the former.

Reply:
I don't see how it can go negative...`bulkStartTime` is set in `beforeBulk` and the calculation happens in `afterBulk`...